paddle/fluid/operators/nccl_op_test.cu.cc

```diff
@@ -97,7 +108,7 @@ class NCCLTester : public ::testing::Test {
       send_tensor->Resize(kDims);
       send_tensor->mutable_data<T>(kDims, place);
 
-      std::vector<T> send_vector(f::product(kDims), gpu_id);
+      std::vector<T> send_vector(f::product(kDims), GetGPUData(gpu_id));
       paddle::framework::TensorFromVector<T>(send_vector, *ctx, send_tensor);
       ctx->Wait();
```

Comment:
Is it necessary to synchronize here? I think the copying will synchronize the GPU and CPU in line 179.

Reply:
The copying is `cudaMemcpyAsync`. Looks like we need to add a wait to line 179...

https://github.com/PaddlePaddle/Paddle/blob/0d49b9214049eab676c1e8e37cbcd04ed25f9065/paddle/fluid/memory/memcpy.cc#L30-L36

Reply:
I don't think so because the memory is pageable in CPU side, the copying doesn't return immediately until the copy has completed.
The current CUDA Runtime Documentation [states](http://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-sync):
**Asynchronous(Memcpy): 
     - For transfers from device memory to pageable host memory, the function will return only once the copy has completed.**

Reply:
A reference https://devtalk.nvidia.com/default/topic/899020/does-cudamemcpyasync-require-pinned-memory-/

Reply:
:+1: